### PR TITLE
make ADTelemetry thread-safe

### PR DESCRIPTION
**Pull request by jasoncoolmax**

### CODE REVIEW

**Comment by RPangrle** [comment]
Do not check in directly to master, you need to create a hotfix branch first.

File: ADAL/src/telemetry/ADTelemetry.m [hunk 1]
```diff
@@ -107,39 +107,45 @@ - (NSString*)registerNewRequest
...
+    @synchronized(self)
     {
-        return;
+        if ([NSString adIsStringNilOrBlank:requestId] || [NSString adIsStringNilOrBlank:eventName])
+        {
+            return;
+        }
+        
+        NSDate* currentTime = [NSDate date];
+        [_eventTracking setObject:currentTime
```
Comment (RPangrle) [inline ADAL/src/telemetry/ADTelemetry.m]: The setObject call is the only thing that needs to be in the synchronized block. Try to keep your critical sections as small as possible

File: ADAL/src/telemetry/ADTelemetry.m [hunk 2]
```diff
@@ -107,39 +107,45 @@ - (NSString*)registerNewRequest
...
+        NSString* key = [self getEventTrackingKey:requestId eventName:eventName];
+        NSDate* startTime = [_eventTracking objectForKey:key];
+        if (!startTime)
+        {
+            return;
+        }
+        [event setStartTime:startTime];
+        [event setStopTime:stopTime];
+        [event setResponseTime:[stopTime timeIntervalSinceDate:startTime]];
+        [_eventTracking removeObjectForKey:key];
```
Comment (RPangrle) [inline ADAL/src/telemetry/ADTelemetry.m]: Same as above. The only part of this that needs to be protected is the part mutating the dictionary

Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADTelemetry.m]: addressed :)

**Comment by RPangrle** [comment]
Branch should be named hotfix/2.3.1

**Comment by jasoncoolmax** [comment]
Sure

File: ADAL/src/telemetry/ADTelemetry.m [hunk 3]
```diff
@@ -129,17 +132,21 @@ - (void)stopEvent:(NSString*)requestId
...
+        
+        [_dispatcher receive:requestId event:event];
```
Comment (RPangrle) [inline ADAL/src/telemetry/ADTelemetry.m]: Please pull this dispatcher call out of the syncronized block.

**Pull request merged**